src/app/metadata-schema-form/metadata-form/metadata-form.component.spec.ts

```diff
@@ -1,104 +1,197 @@
+      config: config as MetadataFormConfig,
+      metadataRegistry: metadataRegistrySpy,
+      formGroup: formGroupSpy
+    } as MetadataForm;
+    metadataFormSvc.createForm.and.returnValue(form);
+    component.form = form;
+    component.formGroup = formGroupSpy;
+    component.schema = schema;
+    component.config = config;
+  }
+
+  beforeEach(() => {
+    metadataFormSvc = jasmine.createSpyObj('MetadataFormService', ['createForm', 'cleanFormData']);
+    metadataRegistrySpy = jasmine.createSpyObj('MetadataRegistry', ['buildMetadataRegistry']);
+    formGroupSpy = jasmine.createSpyObj('FormGroup', ['getRawValue', 'valueChanges'], ['valid']);
     TestBed.configureTestingModule({
       providers: [
-        { provide: MetadataFormService, useValue: metadataFormSvc},
+        {provide: MetadataFormService, useValue: metadataFormSvc},
       ],
       declarations: [MetadataFormComponent]
-    })
-      .compileComponents();
-  }));
-
-  beforeEach(() => {
+    }).compileComponents();
     fixture = TestBed.createComponent(MetadataFormComponent);
     component = fixture.componentInstance;
-    component.schema = getSchema();
-    component.data = getData();
-    component.config = createMetadataFormConfig(tabs);
-    component.selectedTabKey = 'content';
-    component.visibleTabs = tabs;
-    fixture.detectChanges();
+    spyOn(component, 'onFormChange');
   });
 
   it('should create', () => {
     expect(component).toBeTruthy();
   });
 
-  it('tabIsVisible should pass with null content', () => {
-    expect(component.tabIsVisible(contentTab)).toBe(false);
+  describe('tab tests', () => {
+    const contentTab = createTab('content.contributors');
+    const adminTab = createTab('admin');
+    const tabs = [contentTab, adminTab];
+
+    const config = {
+      viewMode: true,
+      hideEmptyFields: true,
+      layout: {
+        tabs: tabs
+      }
+    } as MetadataFormConfig;
+
+    const data = {
+      content: null,
+      admin: {
+        key: 'value'
+      }
+    };
+
+    beforeEach(() => {
+      configureFormAndSvc({config: config});
+      formGroupSpy.getRawValue.and.returnValue(data);
+      component.data = data;
+      component.selectedTabKey = 'content';
+      component.visibleTabs = tabs;
+      fixture.detectChanges();
+    });
+
+    it('tabIsVisible should pass with null content', () => {
+      expect(component.tabIsVisible(contentTab)).toBe(false);
+    });
+  });
+
+  describe('cleanFormData tests', () => {
+    const cleanTest = {
+      unsetAttribute: null,
+      setAttribute: 'value',
+      unsetObject: {},
+      setObject: {
+        unsetObjectAttribute: null,
+        setObjectAttribute: 'objectValue'
+      }
+    };
+
+    beforeEach(() => {
+      formGroupSpy.getRawValue.and.returnValue(cleanTest);
+      component.data = cleanTest;
+    });
+
+    const defaultBehaviourTest = test => {
+      it(`cleanFields ${test.name} (default) behaviour: should send all data to be cleaned`, () => {
```

Comment:
I think the `...` should be changed to a normal apostrophe, like this one: '...'.
The same issue exists in the next `it` section.

Reply:
The difference is between a string literal and a parameterised / interpolated string
```
let string = 'hello world'
let string = `hello $(user_name)`
```

Also, I have a new commit that should help the test names make more sense.

Reply:
Thanks for summarising it to me. 👍